### PR TITLE
CRC24 performance improvement (32 bits in parallel)

### PR DESCRIPTION
**Pull request by flowher**
This is a contribution from Ribose Inc (@riboseinc).

    Algorithm uses 4 tables of precalculated CRC24 values, thanks to which
    it can process in parallel 32 bits of data. This tric doubles performance
    Further improvements are possible.

    Results - (tested with RNP) processing 1GB armor data

    ```
    OLD: rnp --enarmor=msg /tmp/1gb.rnd --output  4.48s user 0.89s system 98% cpu 5.429 total
    NEW: rnp --enarmor=msg /tmp/1gb.rnd --output  2.38s user 0.86s system 79% cpu 4.089 total

    OLD: rnp --dearmor out.xxx --output out.d  5.58s user 0.65s system 98% cpu 6.338 total
    NEW: rnp --dearmor out.xxx --output out.d  3.28s user 0.84s system 96% cpu 4.275 total
    ```

Please let me know if something needs to be changed in the review.

### CODE REVIEW

File: src/lib/hash/checksum/crc24/crc24.cpp [hunk 1]
```diff
@@ -20,87 +195,34 @@ std::unique_ptr<HashFunction> CRC24::copy_state() const
...
-      tmp = TABLE[((tmp >> 16) ^ input[ 7]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[ 8]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[ 9]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[10]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[11]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[12]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[13]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[14]) & 0xFF] ^ (tmp << 8);
-      tmp = TABLE[((tmp >> 16) ^ input[15]) & 0xFF] ^ (tmp << 8);
+      load_le(d, input, 16);
```
Comment (randombit) [inline src/lib/hash/checksum/crc24/crc24.cpp]: This should be 4; the parameter here is the number of words being loaded, not the number of bytes.

File: src/lib/hash/checksum/crc24/crc24.cpp [hunk 2]
```diff
@@ -20,87 +195,34 @@ std::unique_ptr<HashFunction> CRC24::copy_state() const
...
 
 /*
 * Finalize a CRC24 Checksum
 */
 void CRC24::final_result(uint8_t output[])
    {
-   for(size_t i = 0; i != 3; ++i)
-      output[i] = get_byte(i+1, m_crc);
+   for(int i=0; i<3; i++)
+      output[i] = get_byte(3-i, m_crc);
```
Comment (randombit) [inline src/lib/hash/checksum/crc24/crc24.cpp]: It's probably simpler to just unroll this loop
output[0] = get_byte(3, m_crc);
...

File: src/lib/hash/checksum/crc24/crc24.cpp [hunk 3]
```diff
@@ -20,87 +195,34 @@ std::unique_ptr<HashFunction> CRC24::copy_state() const
...
-      0x0057182A, 0x00D154D1, 0x0026359F, 0x00A07964, 0x00ACE092, 0x002AAC69,
-      0x00B5D37E, 0x00339F85, 0x003F0673, 0x00B94A88, 0x0087B4A6, 0x0001F85D,
-      0x000D61AB, 0x008B2D50, 0x00145247, 0x00921EBC, 0x009E874A, 0x0018CBB1,
-      0x00E37B16, 0x006537ED, 0x0069AE1B, 0x00EFE2E0, 0x00709DF7, 0x00F6D10C,
-      0x00FA48FA, 0x007C0401, 0x0042FA2F, 0x00C4B6D4, 0x00C82F22, 0x004E63D9,
-      0x00D11CCE, 0x00575035, 0x005BC9C3, 0x00DD8538 };
-
+   uint32_t d[4];
    uint32_t tmp = m_crc;
+
```
Comment (randombit) [inline src/lib/hash/checksum/crc24/crc24.cpp]: Does it make sense to first consume bytes until the input is word aligned, then read 16 at a time? Probably doesn't matter for x86 but may for other CPUs where misaligned loads can be expensive.

**Comment by randombit** [comment]
Nice, thanks. Can you add a copyright statement at the top of the file? The tests are crashing in CI, presumably the overwrite I commented on. Also if you could add a brief comment at the top explaining the theory of operation that would be nice. Just 1 or 2 lines is fine. Basically T1[i] is (2^8 * T0[i]) in GF(2^24) modulo the CRC24 polynomial, right?

File: src/lib/hash/checksum/crc24/crc24.cpp (hunk 3 above)
Comment (flowher) [reply src/lib/hash/checksum/crc24/crc24.cpp]: I think it makes sense.

File: src/lib/hash/checksum/crc24/crc24.cpp (hunk 1 above)
Comment (flowher) [reply src/lib/hash/checksum/crc24/crc24.cpp]: argh... yes of course. didn't failed locally, so I havn't noticed

**Comment by flowher** [comment]
Okis, adding it now. Yes that's the basic idea.

**Comment by flowher** [comment]
Done. Let me know if you something else. And thanks for review

**Comment by flowher** [comment]
Errors seem unrelated

```
SM2/SM3 decryption ran 10 tests in 187.50 msec 2 FAILED
Failure 1: SM2/SM3 decryption: incorrect ciphertext different produced matching
Failure 2: SM2/SM3 decryption unexpected result for used corrupted ciphertext
```

**Comment by ronaldtse** [comment]
@flowher @randombit the SM2/SM3 issue may be due to the changed ASN.1 output of SM2?

**Comment by flowher** [comment]
Checking...

**Comment by flowher** [comment]
Please don't merge it even if it's OK. Final merge must be done from riboseinc account. It's not possible to change source branch, so after everything is OK I'll open new PR with same code from ribose branch

**Comment by randombit** [comment]
Looks good to me. SM2 problem is unrelated. Likely we are flipping a bit in some region that gets ignored during decryption. I'll check it out.

**Comment by flowher** [comment]
Okis, then I'm closing this PR and will open exact same from ribose

**Pull request closed without merging**